Use `__cpp_aligned_new` instead of hand-rolling the implementation

tamird opened the pull request:
The current feature detection seems to be incorrect, resulting in fallback to
the manual implementation, which in turn returns untagged pointers under
hwasan.
Use new expressions instead to delegate the work to the C++ compiler.

Differential Revision: D75806635

swolchok reviewed (commented):
Do we know whether this actually works on Apple platforms? https://github.com/pytorch/executorch/pull/10660 (the PR that added the previous implementation) cited some concerns about aligned_alloc.

I'm inclined to accept this on the grounds that we should presume that standard library functions that exist on a platform actually work until proven otherwise, but I'd like to let @lucylq chime in.

tamird commented:
> Do we know whether this actually works on Apple platforms? #10660 (the PR that added the previous implementation) cited some concerns about aligned_alloc.
> 
> I'm inclined to accept this on the grounds that we should presume that standard library functions that exist on a platform actually work until proven otherwise, but I'd like to let @lucylq chime in.

We have tests on apple platforms, so that should be sufficient to prove that it works. Yes?

lucylq reviewed (approved):
thanks for the changes!

tamird commented:
@swolchok this keeps failing to land with: 

> The following GitHub Pull Request(s) are not mergeable on GitHub: https://github.com/pytorch/executorch/pull/11293. This Diff cannot land if the Pull Request cannot be merged. Please rebase the diff and re-export it to GitHub or resolve the merge conflict on GitHub and import the changes.

Yet this PR shows no merge conflicts. This has persisted through several rebases. Do you know what's going on?

swolchok commented:
> @swolchok this keeps failing to land with:
> 
> > The following GitHub Pull Request(s) are not mergeable on GitHub: #11293. This Diff cannot land if the Pull Request cannot be merged. Please rebase the diff and re-export it to GitHub or resolve the merge conflict on GitHub and import the changes.
> 
> Yet this PR shows no merge conflicts. This has persisted through several rebases. Do you know what's going on?

how are you updating the PR? if you sent it out internally, you need to update the internal diff and export that diff to github; I see force pushes which looks like maybe you are updating the PR through some separate mechanism like working with github directly.

tamird commented:
> > @swolchok this keeps failing to land with:
> > > The following GitHub Pull Request(s) are not mergeable on GitHub: #11293. This Diff cannot land if the Pull Request cannot be merged. Please rebase the diff and re-export it to GitHub or resolve the merge conflict on GitHub and import the changes.
> > 
> > 
> > Yet this PR shows no merge conflicts. This has persisted through several rebases. Do you know what's going on?
> 
> how are you updating the PR? if you sent it out internally, you need to update the internal diff and export that diff to github; I see force pushes which looks like maybe you are updating the PR through some separate mechanism like working with github directly.

No, those force pushes are done by phabricator.

It seems that PRs that have both internal and external changes are broken because the `mergeable_state` is always `"unstable"`.

Commit:
Use `__cpp_aligned_new` instead of hand-rolling the implementation (#11293)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/11293

The current feature detection seems to be incorrect, resulting in fallback to
the manual implementation, which in turn returns untagged pointers under
hwasan.

Use new expressions instead to delegate the work to the C++ compiler.

Reviewed By: lucylq

Differential Revision: D75806635
